src/coroutine.rs

```diff
@@ -0,0 +1,218 @@
+//! Python coroutine implementation, used notably when wrapping `async fn`
+//! with `#[pyfunction]`/`#[pymethods]`.
+use crate::exceptions::{PyRuntimeError, PyStopIteration};
+use crate::pyclass::IterNextOutput;
+use crate::sync::GILOnceCell;
+use crate::types::PyIterator;
+use crate::{intern, IntoPy, Py, PyAny, PyErr, PyObject, PyResult, Python};
+use futures::task::ArcWake;
+use pyo3_macros::{pyclass, pymethods};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+const COROUTINE_REUSED_ERROR: &str = "cannot reuse already awaited coroutine";
+
+/// Python coroutine wrapping a [`Future`].
+#[pyclass(crate = "crate")]
+pub struct Coroutine {
+    future: Option<Pin<Box<dyn Future<Output = PyResult<PyObject>> + Send>>>,
+    #[allow(clippy::type_complexity)]
```

Comment:
I think type aliases would be nice, e.g. `PyFuture` and `PyThrowCallback`.

Reply:
`PyFuture` can be ambiguous, because others `PyXxx` (`PyDict`, `PyIterator`, etc.) are Python objects, and there is a `Future` type in Python (there are two in fact).
Strangely, clippy only warns about type complexity for `Option<Box<dyn FnMut(&PyAny) + Send>>`, not for the future.
In pyo3-async, I've named it `ThrowCallback`, I could reuse the same name here.

Reply:
I am not keen on any of the names and will defer to your experience in building this, I just think any type aliases will be an improvement.